### ISSUE TITLE
save_topk_models() timeout in when num_trainer > 0

### ISSUE THREAD

**Issue by CongWeilin** [awslabs/graphstorm#276]
I get this error when using `num_trainer > 1`. It seems like a multi-process dead-lock issue according to `Timed out waiting 1800000ms for recv operation to complet`?

Full error:

```
Traceback (most recent call last):
  File "main_ssl.py", line 156, in <module>
    main(args)
  File "main_ssl.py", line 134, in main
    trainer.fit(
  File "/home/ubuntu/workspace/ssl_utils.py", line 207, in fit
    early_stop = True
  File "/usr/local/lib/python3.8/dist-packages/graphstorm-0.1.0.post1-py3.8.egg/graphstorm/trainer/gsgnn_trainer.py", line 263, in save_topk_models
Traceback (most recent call last):
  File "main_ssl.py", line 156, in <module>
      File "/usr/local/lib/python3.8/dist-packages/graphstorm-0.1.0.post1-py3.8.egg/graphstorm/trainer/gsgnn_trainer.py", line 190, in save_model
main(args)
  File "main_ssl.py", line 134, in main
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2791, in barrier
    trainer.fit(
  File "/home/ubuntu/workspace/ssl_utils.py", line 147, in fit
    work.wait()
RuntimeError: [../third_party/gloo/gloo/transport/tcp/unbound_buffer.cc:81] Timed out waiting 1800000ms for recv operation to complete
    loss.backward()
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/_tensor.py", line 396, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/autograd/__init__.py", line 173, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: [../third_party/gloo/gloo/transport/tcp/pair.cc:598] Connection closed by peer [172.31.5.112]:33476
Client[4] in group[0] is exiting...
Client[7] in group[0] is exiting...
```

**Comment by CongWeilin** [comment]
I got the above error by using single machine standalone mode with `num_trainer=2` and `num_sampler=8`, with `torch=1.12.0`.

**Comment by CongWeilin** [comment]
I am still having this error when using `graphstorm=0.2` version. It seems like the code was stopped at `th.distributed.barrier()` https://github.com/awslabs/graphstorm/blob/d010ed45d549cf4a3029854053f6dc0354ccf2aa/python/graphstorm/trainer/gsgnn_trainer.py#L190C1-L191C1 to wait for the sparse optimizer to update its parameters? Its not a storage issue as I still have more than `1.2 T` available. 

My implementation is based on `GSgnnLinkPredictionTrainer` and its save_model part is using existing `graphstorm`'s functions. The code can successfully save models for already 28 rounds, therefore it should be a RAM OOM issue ... To reproduce this error, one can just take the MAG link prediction example, adding learnable node embedding to all nodes, and frequently save model params. 

```
Epoch 00000 | Batch 14460 | Train Loss: 0.2059 | Time: 0.8983
Epoch 00000 | Batch 14480 | Train Loss: 0.1833 | Time: 0.9150
Traceback (most recent call last):
  File "main_ssl.py", line 90, in <module>
Traceback (most recent call last):
  File "main_ssl.py", line 90, in <module>
    main(args)
  File "main_ssl.py", line 74, in main
    trainer.fit(
  File "/home/ubuntu/workspace/step_by_step_dev/trainer.py", line 86, in fit
    main(args)
  File "main_ssl.py", line 74, in main
    trainer.fit(
  File "/home/ubuntu/workspace/step_by_step_dev/trainer.py", line 108, in fit
    early_stop = self.save_topk_models(
  File "/home/ubuntu/workspace/step_by_step_dev/trainer.py", line 168, in save_topk_models
    self.optimizer.step()
  File "/home/ubuntu/pytorch-1-12-0/lib/python3.8/site-packages/graphstorm-0.2-py3.8.egg/graphstorm/model/gnn.py", line 100, in step
    self.save_model(model, epoch, i, save_model_path)
  File "/home/ubuntu/pytorch-1-12-0/lib/python3.8/site-packages/graphstorm-0.2-py3.8.egg/graphstorm/trainer/gsgnn_trainer.py", line 190, in save_model
  File "/home/ubuntu/pytorch-1-12-0/lib/python3.8/site-packages/dgl/distributed/optim/pytorch/sparse_optim.py", line 343, in step
  File "/home/ubuntu/pytorch-1-12-0/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2791, in barrier
    alltoall_cpu(
  File "/home/ubuntu/pytorch-1-12-0/lib/python3.8/site-packages/dgl/distributed/optim/pytorch/utils.py", line 26, in alltoall_cpu
    dist.scatter(
  File "/home/ubuntu/pytorch-1-12-0/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2359, in scatter
    work.wait()
RuntimeError: [../third_party/gloo/gloo/transport/tcp/unbound_buffer.cc:81] Timed out waiting 1800000ms for recv operation to complete
Client[3] in group[0] is exiting...
    work.wait()
RuntimeError: [../third_party/gloo/gloo/transport/tcp/pair.cc:598] Connection closed by peer [172.31.11.219]:53988
Client[0] in group[0] is exiting...
```
